relay-filter/src/interface.rs

```diff
@@ -0,0 +1,74 @@
+//! This module contains the trait for items that can be filtered by Inbound Filters, plus
+//! the implementation for [`Event`].
+use url::Url;
+
+use relay_event_schema::protocol::{Csp, Event, EventType, Exception, LogEntry, Values};
+
+/// A data item to which filters can be applied.
+pub trait Filterable {
+    /// The CSP report contained in the item. Only for CSP reports.
+    fn csp(&self) -> Option<&Csp>;
+
+    /// The exception values of the item. Only for error events.
+    fn exceptions(&self) -> Option<&Values<Exception>>;
+
+    /// The IP address of the client that sent the data.
+    fn ip_addr(&self) -> Option<&str>;
+
+    /// The logentry message. Only for error events.
+    fn logentry(&self) -> Option<&LogEntry>;
+
+    /// The release string of the data item.
+    fn release(&self) -> Option<&str>;
+
+    /// The transaction name. Only for transaction events.
+    fn transaction(&self) -> Option<&str>;
+
+    /// The URL from which the data originates. Used for localhost filtering.
+    fn url(&self) -> Option<Url>;
+
+    /// The user agent of the client that sent the data.
+    fn user_agent(&self) -> Option<&str>;
+}
+
+impl Filterable for Event {
+    fn csp(&self) -> Option<&Csp> {
+        if self.ty.value() != Some(&EventType::Csp) {
+            return None;
+        }
+        self.csp.value()
+    }
+
+    fn exceptions(&self) -> Option<&Values<Exception>> {
+        self.exceptions.value()
+    }
+
+    fn ip_addr(&self) -> Option<&str> {
+        let user = self.user.value()?;
+        Some(user.ip_address.value()?.as_ref())
```

Comment:
There's a getter implementation for this, and we already import the trait in the crate, so this should be possible?
```suggestion
        get_value!(self.user.ip_address!).as_ref()
```

Reply:
Good idea, but `get_value!` only works if the root object is `Annotated` AFAIK. We could change that, but it seems out of scope. Side note: `ip_address!` unwraps the value, the exclamation mark in `get_value!` paths should only be used in tests.